src/test/java/bio/terra/controller/DatasetOperationTest.java

```diff
@@ -0,0 +1,260 @@
+        studyRequest.setName(randomizedName(studyRequest.getName(), ""));
+        createTestStudy();
+
+        String datasetJson = IOUtils.toString(classLoader.getResourceAsStream("dataset-test-dataset.json"));
+        datasetRequest = objectMapper.readerFor(DatasetRequestModel.class).readValue(datasetJson);
+        datasetRequest.getContents().get(0).getSource().setStudyName(studyRequest.getName());
+    }
+
+    // TODO: add @After for study delete when that is hooked up
+
+    @Test
+    public void testHappyPath() throws Exception {
+        DatasetSummaryModel summaryModel = createTestDataset(datasetRequest, "happy");
+
+        DatasetModel datasetModel = getTestDataset(summaryModel.getId());
+
+        deleteTestDataset(datasetModel.getId());
+        // Duplicate delete should work
+        deleteTestDataset(datasetModel.getId());
+
+        getNonexistentDataset(datasetModel.getId());
+    }
+
+    @Test
+    public void testEnumeration() throws Exception {
+        // Unit tests exercise the array bounds, so here we don't fuss with that here.
+        // Just make sure we get the same dataset summary that we made.
+        List<DatasetSummaryModel> datasetList = new ArrayList<>();
+        for (int i = 0; i < 5; i++) {
+            DatasetSummaryModel dataset = createTestDataset(datasetRequest, "_enum_");
+            datasetList.add(dataset);
+        }
+
+        DatasetSummaryModel[] enumeratedArray = enumerateTestDatasets();
+
+        // The enumeratedArray may contain more datasets than just the set we created,
+        // but ours should be in order in the enumeration. So we do a merge waiting until we match
+        // by id and then comparing contents.
+        int compareIndex = 0;
+        for (int i = 0; i < enumeratedArray.length; i++) {
+            if (enumeratedArray[i].getId().equals(datasetList.get(compareIndex).getId())) {
+                assertThat("Enumeration summary matches create summary",
+                        enumeratedArray[i], equalTo(datasetList.get(compareIndex)));
+                compareIndex++;
+            }
+        }
+        assertThat("we found all datasets", compareIndex, equalTo(5));
+
+        for (int i = 0; i < 5; i++) {
+            deleteTestDataset(enumeratedArray[i].getId());
+        }
+    }
+
+    private DatasetSummaryModel createTestDataset(DatasetRequestModel datasetRequest, String infix) throws Exception {
+        String baseName = datasetRequest.getName();
+        String datasetName = randomizedName(baseName, infix);
+        datasetRequest.setName(datasetName);
+
+        String jsonRequest = objectMapper.writeValueAsString(datasetRequest);
+
+        MvcResult result = mvc.perform(post("/api/repository/v1/datasets")
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(jsonRequest))
+// TODO: swagger field validation errors do not set content type; they log and return nothing
```

Comment:
I added [DR-157](https://broadinstitute.atlassian.net/jira/software/projects/DR/boards/517/backlog?selectedIssue=DR-157) to track.